From e88c0f7802cba2076ded6569dd70e24f545d36da Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Nicolai=20H=C3=A4hnle?= <nicolai.haehnle@amd.com>
Date: Wed, 12 Jul 2017 15:38:57 +0200
Subject: [PATCH 11/14] radeonsi/gfx9: fix crash building monolithic merged
 ES-GS shader

Forwarding from the ES prolog to the ES just barely exceeds the current
maximum array size when 16 vertex attributes are used. Give it a decent
bump to account for merged shaders having up to 32 user SGPRs.

Fixes a crash in GL45-CTS.multi_bind.draw_bind_vertex_buffers.

Cc: mesa-stable@lists.freedesktop.org
---
 src/gallium/drivers/radeonsi/si_shader.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/src/gallium/drivers/radeonsi/si_shader.c b/src/gallium/drivers/radeonsi/si_shader.c
index 2c92269..6c88dbe 100644
--- a/src/gallium/drivers/radeonsi/si_shader.c
+++ b/src/gallium/drivers/radeonsi/si_shader.c
@@ -5852,9 +5852,12 @@ static void si_build_wrapper_function(struct si_shader_context *ctx,
 {
 	struct gallivm_state *gallivm = &ctx->gallivm;
 	LLVMBuilderRef builder = ctx->gallivm.builder;
-	/* PS epilog has one arg per color component */
-	LLVMTypeRef param_types[48];
-	LLVMValueRef initial[48], out[48];
+
+        /* PS epilog has one arg per color component; gfx9 merged shader
+         * prologs need to forward 32 user SGPRs.
+         */
+        LLVMTypeRef param_types[64];
+        LLVMValueRef initial[64], out[64];
 	LLVMTypeRef function_type;
 	unsigned num_params;
 	unsigned num_out, initial_num_out;
@@ -6074,6 +6077,7 @@ static void si_build_wrapper_function(struct si_shader_context *ctx,
 				LLVMValueRef val =
 					LLVMBuildExtractValue(builder, ret, i, "");
 
+				assert(num_out < ARRAY_SIZE(out));
 				out[num_out++] = val;
 
 				if (LLVMTypeOf(val) == ctx->i32) {
-- 
2.7.4

